From 13b8199f0a47a3874366f884c7255562141dad19 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Sun, 9 Oct 2022 19:47:38 +0800
Subject: [PATCH 0904/1076] [Huawei] cgroup: support cgroup writeback on
 cgroupv1

Offering: HULK
hulkl inclusion
category: feature
bugzilla: 187773

-------------------------------

In cgroupv1, cgroup writeback is not supproted for two problems:
1) Blkcg_css and memcg_css are mounted on different cgroup trees.
   Therefore, blkcg_css cannot be found according to a certain memcg_css.
2) Buffer I/O is worked by kthread, which is in the root_blkcg.
   Therefore, blkcg cannot limit wbps and wiops of buffer I/O.

We solve the two problems to support cgroup writeback on cgroupv1.
1) Blkcg_root_css is attached to a memcg when is was created.
2) We add a number "wb_blkio_ino" in mem_cgroup_legacy_files.
   User can attach a blkcg_css to a memcg through echo the file
   inode of the blkcg into the wb_blkio of the memcg.
3) inode_cgwb_enabled() return true when memcg and io are mounted on
   cgroupv2 or cgroupv1.
4) Buffer I/O can find a blkcg according to its page's memcg.

Thus, a memcg can find a certain blkcg, and cgroup writeback can be
supported on cgroupv1.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 block/blk-cgroup.c          |   1 +
 include/linux/backing-dev.h |  14 +++--
 include/linux/blk-cgroup.h  |   5 +-
 include/linux/memcontrol.h  |   5 ++
 kernel/cgroup/cgroup.c      |   7 +++
 mm/backing-dev.c            | 102 +++++++++++++++++++++++++++++++++++-
 mm/memcontrol.c             |  84 ++++++++++++++++++++++++++++-
 7 files changed, 210 insertions(+), 8 deletions(-)

diff --git a/block/blk-cgroup.c b/block/blk-cgroup.c
index bd13079d63f9..dd1193494016 100644
--- a/block/blk-cgroup.c
+++ b/block/blk-cgroup.c
@@ -1146,6 +1146,7 @@ blkcg_css_alloc(struct cgroup_subsys_state *parent_css)
 	INIT_HLIST_HEAD(&blkcg->blkg_list);
 #ifdef CONFIG_CGROUP_WRITEBACK
 	INIT_LIST_HEAD(&blkcg->cgwb_list);
+	INIT_LIST_HEAD(&blkcg->memcg_list);
 #endif
 	list_add_tail(&blkcg->all_blkcgs_node, &all_blkcgs);
 
diff --git a/include/linux/backing-dev.h b/include/linux/backing-dev.h
index 44df4fcef65c..61aed9511150 100644
--- a/include/linux/backing-dev.h
+++ b/include/linux/backing-dev.h
@@ -178,6 +178,11 @@ void wb_memcg_offline(struct mem_cgroup *memcg);
 void wb_blkcg_offline(struct blkcg *blkcg);
 int inode_congested(struct inode *inode, int cong_bits);
 
+void wb_kill_memcg_blkcg(struct cgroup_subsys_state *css);
+void wb_attach_memcg_to_blkcg(struct cgroup_subsys_state *memcg_css,
+			      struct cgroup_subsys_state *blkcg_css);
+bool cgroup1_writeback_enabled(void);
+
 /**
  * inode_cgwb_enabled - test whether cgroup writeback is enabled on an inode
  * @inode: inode of interest
@@ -193,10 +198,11 @@ static inline bool inode_cgwb_enabled(struct inode *inode)
 {
 	struct backing_dev_info *bdi = inode_to_bdi(inode);
 
-	return cgroup_subsys_on_dfl(memory_cgrp_subsys) &&
-		cgroup_subsys_on_dfl(io_cgrp_subsys) &&
-		(bdi->capabilities & BDI_CAP_WRITEBACK) &&
-		(inode->i_sb->s_iflags & SB_I_CGROUPWB);
+	return ((cgroup_subsys_on_dfl(memory_cgrp_subsys) &&
+	       cgroup_subsys_on_dfl(io_cgrp_subsys)) ||
+	       cgroup1_writeback_enabled()) &&
+	       (bdi->capabilities & BDI_CAP_WRITEBACK) &&
+	       (inode->i_sb->s_iflags & SB_I_CGROUPWB);
 }
 
 /**
diff --git a/include/linux/blk-cgroup.h b/include/linux/blk-cgroup.h
index 994ff06de40f..4c1875d69c36 100644
--- a/include/linux/blk-cgroup.h
+++ b/include/linux/blk-cgroup.h
@@ -59,8 +59,11 @@ struct blkcg {
 #ifdef CONFIG_CGROUP_WRITEBACK
 	struct list_head		cgwb_list;
 #endif
-
+#if defined(CONFIG_CGROUP_WRITEBACK) && !defined(__GENKSYMS__)
+	struct list_head		memcg_list;
+#else
 	KABI_RESERVE(1)
+#endif
 	KABI_RESERVE(2)
 	KABI_RESERVE(3)
 	KABI_RESERVE(4)
diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 3a4ea9ed39fd..a6483f69768f 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -375,8 +375,13 @@ struct mem_cgroup {
 #ifdef CONFIG_DYNAMIC_HUGETLB
 	struct dhugetlb_pool *hpool;
 #endif
+#if defined(CONFIG_CGROUP_WRITEBACK) && !defined(__GENKSYMS__)
+	struct cgroup_subsys_state *wb_blk_css;
+	struct list_head memcg_node;
+#else
 	KABI_RESERVE(1)
 	KABI_RESERVE(2)
+#endif
 	KABI_RESERVE(3)
 	KABI_RESERVE(4)
 	KABI_RESERVE(5)
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 6a7d58d7c57a..25fe37e6d21f 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -60,6 +60,10 @@
 #include <linux/dynamic_hugetlb.h>
 #include <net/sock.h>
 
+#ifdef CONFIG_CGROUP_WRITEBACK
+#include <linux/backing-dev.h>
+#endif
+
 #define CREATE_TRACE_POINTS
 #include <trace/events/cgroup.h>
 
@@ -5636,6 +5640,9 @@ static void kill_css(struct cgroup_subsys_state *css)
 	 */
 	css_get(css);
 
+#ifdef CONFIG_CGROUP_WRITEBACK
+	wb_kill_memcg_blkcg(css);
+#endif
 	/*
 	 * cgroup core guarantees that, by the time ->css_offline() is
 	 * invoked, no new css reference will be given out via
diff --git a/mm/backing-dev.c b/mm/backing-dev.c
index ca770a783a9f..cbdf63ec2c32 100644
--- a/mm/backing-dev.c
+++ b/mm/backing-dev.c
@@ -424,6 +424,21 @@ static void cgwb_remove_from_bdi_list(struct bdi_writeback *wb)
 	spin_unlock_irq(&cgwb_lock);
 }
 
+static struct cgroup_subsys_state *cgwbv1_get_blkcss(struct mem_cgroup *memcg)
+{
+	struct cgroup_subsys_state *blkcg_css;
+
+	rcu_read_lock();
+	blkcg_css = memcg->wb_blk_css;
+	if (!css_tryget_online(blkcg_css)) {
+		blkcg_css = blkcg_root_css;
+		css_get(blkcg_css);
+	}
+	rcu_read_unlock();
+
+	return blkcg_css;
+}
+
 static int cgwb_create(struct backing_dev_info *bdi,
 		       struct cgroup_subsys_state *memcg_css, gfp_t gfp)
 {
@@ -436,7 +451,11 @@ static int cgwb_create(struct backing_dev_info *bdi,
 	int ret = 0;
 
 	memcg = mem_cgroup_from_css(memcg_css);
-	blkcg_css = cgroup_get_e_css(memcg_css->cgroup, &io_cgrp_subsys);
+	if (cgroup1_writeback_enabled())
+		blkcg_css = cgwbv1_get_blkcss(memcg);
+	else
+		blkcg_css = cgroup_get_e_css(memcg_css->cgroup,
+					     &io_cgrp_subsys);
 	blkcg = css_to_blkcg(blkcg_css);
 	memcg_cgwb_list = &memcg->cgwb_list;
 	blkcg_cgwb_list = &blkcg->cgwb_list;
@@ -553,9 +572,14 @@ struct bdi_writeback *wb_get_lookup(struct backing_dev_info *bdi,
 	wb = radix_tree_lookup(&bdi->cgwb_tree, memcg_css->id);
 	if (wb) {
 		struct cgroup_subsys_state *blkcg_css;
+		struct mem_cgroup *memcg = mem_cgroup_from_css(memcg_css);
 
 		/* see whether the blkcg association has changed */
-		blkcg_css = cgroup_get_e_css(memcg_css->cgroup, &io_cgrp_subsys);
+		if (cgroup1_writeback_enabled())
+			blkcg_css = cgwbv1_get_blkcss(memcg);
+		else
+			blkcg_css = cgroup_get_e_css(memcg_css->cgroup,
+						     &io_cgrp_subsys);
 		if (unlikely(wb->blkcg_css != blkcg_css || !wb_tryget(wb)))
 			wb = NULL;
 		css_put(blkcg_css);
@@ -1021,3 +1045,77 @@ long wait_iff_congested(int sync, long timeout)
 	return ret;
 }
 EXPORT_SYMBOL(wait_iff_congested);
+
+#ifdef CONFIG_CGROUP_WRITEBACK
+
+#include "../kernel/cgroup/cgroup-internal.h"
+
+static bool cgroup1_writeback __read_mostly;
+
+bool cgroup1_writeback_enabled(void)
+{
+	return !cgroup_subsys_on_dfl(memory_cgrp_subsys) &&
+	       !cgroup_subsys_on_dfl(io_cgrp_subsys) && cgroup1_writeback;
+}
+
+static void wb_kill_memcg(struct cgroup_subsys_state *memcg_css)
+{
+	struct mem_cgroup *memcg = mem_cgroup_from_css(memcg_css);
+
+	list_del_init(&memcg->memcg_node);
+	css_put(memcg->wb_blk_css);
+}
+
+static void wb_kill_blkcg(struct cgroup_subsys_state *blkcg_css)
+{
+	struct mem_cgroup *memcg, *tmp;
+	struct blkcg *blkcg = css_to_blkcg(blkcg_css);
+	struct blkcg *root_blkcg = css_to_blkcg(blkcg_root_css);
+
+	list_for_each_entry_safe(memcg, tmp, &blkcg->memcg_list, memcg_node) {
+		css_get(blkcg_root_css);
+		memcg->wb_blk_css = blkcg_root_css;
+		list_move(&memcg->memcg_node, &root_blkcg->memcg_list);
+		css_put(blkcg_css);
+	}
+}
+
+void wb_kill_memcg_blkcg(struct cgroup_subsys_state *css)
+{
+	struct cgroup_subsys *ss = css->ss;
+
+	lockdep_assert_held(&cgroup_mutex);
+
+	if (ss->id == io_cgrp_id)
+		wb_kill_blkcg(css);
+	else if (ss->id == memory_cgrp_id)
+		wb_kill_memcg(css);
+}
+
+void wb_attach_memcg_to_blkcg(struct cgroup_subsys_state *memcg_css,
+			      struct cgroup_subsys_state *blkcg_css)
+{
+	struct mem_cgroup *memcg = mem_cgroup_from_css(memcg_css);
+	struct cgroup_subsys_state *pre_blkcss = memcg->wb_blk_css;
+	struct blkcg *blkcg = css_to_blkcg(blkcg_css);
+
+	lockdep_assert_held(&cgroup_mutex);
+
+	css_get(blkcg_css);
+	memcg->wb_blk_css = blkcg_css;
+	if (pre_blkcss == NULL)
+		list_add(&memcg->memcg_node, &blkcg->memcg_list);
+	else {
+		list_move(&memcg->memcg_node, &blkcg->memcg_list);
+		css_put(pre_blkcss);
+	}
+}
+
+static int __init enable_cgroup1_writeback(char *s)
+{
+	cgroup1_writeback = true;
+
+	return 1;
+}
+__setup("cgroup1_writeback", enable_cgroup1_writeback);
+#endif
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 3bc216a14762..88e1c0dc78f2 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -422,7 +422,8 @@ struct cgroup_subsys_state *mem_cgroup_css_from_page(struct page *page)
 
 	memcg = page_memcg(page);
 
-	if (!memcg || !cgroup_subsys_on_dfl(memory_cgrp_subsys))
+	if (!memcg || !(cgroup_subsys_on_dfl(memory_cgrp_subsys) ||
+	    cgroup1_writeback_enabled()))
 		memcg = root_mem_cgroup;
 
 	return &memcg->css;
@@ -5195,6 +5196,76 @@ static ssize_t memory_reclaim(struct kernfs_open_file *of, char *buf,
 	return nbytes;
 }
 
+#ifdef CONFIG_CGROUP_WRITEBACK
+#include "../kernel/cgroup/cgroup-internal.h"
+
+static int wb_blkio_show(struct seq_file *m, void *v)
+{
+	char *path;
+	ino_t blkcg_id;
+	struct cgroup *blkcg_cgroup;
+	struct cgroup_subsys_state *blkcg_css;
+	struct mem_cgroup *memcg = mem_cgroup_from_seq(m);
+
+	if (!cgroup1_writeback_enabled())
+		return -EOPNOTSUPP;
+
+	path = kzalloc(PATH_MAX, GFP_KERNEL);
+	if (!path)
+		return -ENOMEM;
+
+	mutex_lock(&cgroup_mutex);
+	blkcg_css = memcg->wb_blk_css;
+	blkcg_cgroup = blkcg_css->cgroup;
+	blkcg_id = cgroup_ino(blkcg_cgroup);
+	cgroup_path(blkcg_cgroup, path, PATH_MAX);
+	seq_printf(m, "wb_blkio_path:%s\n", path);
+	seq_printf(m, "wb_blkio_ino:%lu\n", blkcg_id);
+	kfree(path);
+	mutex_unlock(&cgroup_mutex);
+
+	return 0;
+}
+
+static ssize_t wb_blkio_write(struct kernfs_open_file *of, char *buf,
+			      size_t nbytes, loff_t off)
+{
+	int ret = 0;
+	u64 cgrp_id;
+	struct cgroup_root *root;
+	struct cgroup *blk_cgroup;
+	struct cgroup_subsys_state *blkcg_css;
+	struct cgroup_subsys_state *memcg_css = of_css(of);
+
+	if (!cgroup1_writeback_enabled())
+		return -EOPNOTSUPP;
+
+	buf = strstrip(buf);
+	ret = kstrtou64(buf, 0, &cgrp_id);
+	if (ret)
+		return ret;
+
+	mutex_lock(&cgroup_mutex);
+	root = blkcg_root_css->cgroup->root;
+	blk_cgroup = cgroup1_get_from_id(root, cgrp_id);
+	if (!blk_cgroup) {
+		mutex_unlock(&cgroup_mutex);
+		return -EINVAL;
+	}
+	blkcg_css = cgroup_tryget_css(blk_cgroup, &io_cgrp_subsys);
+	if (!blkcg_css)
+		goto out_unlock;
+	wb_attach_memcg_to_blkcg(memcg_css, blkcg_css);
+	css_put(blkcg_css);
+
+out_unlock:
+	cgroup_put(blk_cgroup);
+	mutex_unlock(&cgroup_mutex);
+
+	return ret < 0 ? ret : nbytes;
+}
+#endif
+
 static struct cftype mem_cgroup_legacy_files[] = {
 	{
 		.name = "usage_in_bytes",
@@ -5392,6 +5463,14 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.name = "reclaim",
 		.write = memory_reclaim,
 	},
+#ifdef CONFIG_CGROUP_WRITEBACK
+	{
+		.name = "wb_blkio_ino",
+		.flags = CFTYPE_NOT_ON_ROOT,
+		.seq_show = wb_blkio_show,
+		.write = wb_blkio_write,
+	},
+#endif
 	{ },	/* terminate */
 };
 
@@ -5652,6 +5731,9 @@ mem_cgroup_css_alloc(struct cgroup_subsys_state *parent_css)
 	if (cgroup_subsys_on_dfl(memory_cgrp_subsys) && !cgroup_memory_nosocket)
 		static_branch_inc(&memcg_sockets_enabled_key);
 
+#ifdef CONFIG_CGROUP_WRITEBACK
+	wb_attach_memcg_to_blkcg(&memcg->css, blkcg_root_css);
+#endif
 	return &memcg->css;
 fail:
 	mem_cgroup_id_remove(memcg);
-- 
2.17.1

